scripts/reconnect-test.sh

```diff
@@ -7,22 +7,29 @@ export RUST_LOG=info
 source ./scripts/setup-tests.sh
 ./scripts/start-fed.sh
 
+server4=$(tail -4 $FM_PID_FILE | head -1)
+server3=$(tail -3 $FM_PID_FILE | head -1)
 server2=$(tail -2 $FM_PID_FILE | head -1)
 server1=$(tail -1 $FM_PID_FILE | head -1)
 
 mine_blocks 110
 await_block_sync
 
+# test a peer missing out on epochs and needing to rejoin
 kill $server1
-
 mine_blocks 100
 await_block_sync
-
 mine_blocks 100
 await_block_sync
-
 ./scripts/start-fed.sh
 # FIXME should await a response from all 4 peers instead of this hack
 sleep 5
 kill $server2
 await_block_sync
+
+# now test what happens if consensus needs to be restarted
+kill $server3
+kill $server4
+./scripts/start-fed.sh
+mine_blocks 100
+await_block_sync
```

Comment:
So we're basically just asserting that this doesn't hang forever?

Reply:
If the peers reconnect then they will sync up to the block height, otherwise the responses will be wrong and this blocks or fails.

Reply:
Does this already work with this PR? :open_mouth: Doesn't this require the still online peers to kill their HBBFT and restart it too? Maybe this only works if we don't crash in an epoch but between two. Should be tested under load imo (probably not in CI though) so there's some probability that the killed member was in the middle of an epoch.

Reply:
Yep it works because all peers are on the same epoch.  If `< 2f+1` of the peers are on the same epoch when consensus crashes there is no way to automatically restart.  I'd be curious to see what happens under load, but there's no guarantee everything will restart.

We should eventually move the consensus signing into hbbft itself so all outcomes can be easily verified, then we'd have no issues with restarting and the logic would be much simpler.  I regret not trying it for this issue in the first place, but I think we were concerned with modifying the hbbft code.